mongocrypt/src/ctx.rs

```diff
@@ -0,0 +1,687 @@
+    /// the key vault collection.
+    ///
+    /// If this context was initialized with `CtxBuilder::build_rewrap_many_datakey`,
+    /// then this BSON has the form { "v": [(BSON document), ...] } where each BSON
+    /// document in the array is a document containing a rewrapped datakey to be
+    /// bulk-updated into the key vault collection.
+    pub fn finalize(&mut self) -> Result<&RawDocument> {
+        let bytes = unsafe {
+            let bin = Binary::new();
+            if !sys::mongocrypt_ctx_finalize(*self.inner.borrow(), *bin.native()) {
+                return Err(self.status().as_error());
+            }
+            bin.bytes()?
+        };
+        rawdoc(bytes)
+    }
+}
+
+/// Indicates the state of the `Ctx`. Each state requires
+/// different handling. See [the integration
+/// guide](https://github.com/mongodb/libmongocrypt/blob/master/integrating.md#state-machine)
+/// for information on what to do for each state.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[non_exhaustive]
+pub enum State {
+    NeedMongoCollinfo,
+    NeedMongoMarkings,
+    NeedMongoKeys,
+    NeedKms,
+    NeedKmsCredentials,
+    Ready,
+    Done,
+    Other(sys::mongocrypt_ctx_state_t),
+}
+
+impl State {
+    fn from_native(state: sys::mongocrypt_ctx_state_t) -> Self {
+        match state {
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_MONGO_COLLINFO => {
+                Self::NeedMongoCollinfo
+            }
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_MONGO_MARKINGS => {
+                Self::NeedMongoMarkings
+            }
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_MONGO_KEYS => Self::NeedMongoKeys,
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_KMS => Self::NeedKms,
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS => {
+                Self::NeedKmsCredentials
+            }
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_READY => Self::Ready,
+            sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_DONE => Self::Done,
+            other => Self::Other(other),
+        }
+    }
+}
+
+/// A scope bounding the processing of (potentially multiple) KMS handles.
+pub struct KmsScope<'ctx> {
+    ctx: &'ctx Ctx,
+}
+
+// This is `Iterator`-like but does not impl that because it's encouraged for multiple `KmsCtx` to
+// be retrieved and processed in parallel, as reflected in the `&self` shared reference rather than
+// `Iterator`'s exclusive `next(&mut self)`.
+impl<'ctx> KmsScope<'ctx> {
+    /// Get the next KMS handle.
+    ///
+    /// Multiple KMS handles may be retrieved at once. Drivers may do this to fan
+    /// out multiple concurrent KMS HTTP requests. Feeding multiple KMS requests
+    /// is thread-safe.
+    ///
+    /// If KMS handles are being handled synchronously, the driver can reuse the same
+    /// TLS socket to send HTTP requests and receive responses.
+    pub fn next_kms_ctx(&self) -> Option<KmsCtx> {
```

Comment:
does this do I/O? If so, might be worth noting so that in the driver we can use `spawn_blocking` or whatever to avoid blocking the event loop.

Reply:
Nope, libmongocrypt does no I/O of its own.